src/Workspaces/Core/Portable/Execution/Serializer.cs

```diff
@@ -88,6 +190,49 @@ public T Deserialize<T>(string kind, ObjectReader reader, CancellationToken canc
             }
         }
 
+        private string GetLanguageName(object value)
+        {
+            // for given object, we need to figure out which language the object belong to. 
+            // we can't blindly get language service since that will bring in language specific dlls.
+            foreach (var languageName in _workspaceServices.SupportedLanguages)
+            {
+                IOptionsSerializationService service;
+                if (_lazyLanguageSerializationService.TryGetValue(languageName, out service))
+                {
+                    if (service.Owns(value))
+                    {
+                        return languageName;
+                    }
+
+                    continue;
+                }
+
+                // this should be only reached once per language value actually belong to
+                var mefWorkspaceServices = _workspaceServices as MefWorkspaceServices;
+                if (mefWorkspaceServices != null)
+                {
+                    MefLanguageServices languageServices;
+                    if (!mefWorkspaceServices.TryGetLanguageServices(languageName, out languageServices))
+                    {
+                        // this is a bit fragile since it depends on implementation detail but there is no other way
+                        // to figure out which language a type belong to without loading other languages
```

Comment:
I would say we should just bite the bullet and make them serializable in the compiler layer.  It's not like that will be contentious (so much of the compiler immutable data is already serializable).  And it will mean we don't have to have really scary code liek this that is so dependent on disparate parts of the system knowing about internal implementation details.